### PR TITLE
[helm] Parameterized all service types

### PR DESCRIPTION
**Pull request by iridian-ks**
Hello Gloo team,

Thanks for all you do!

We have cases where we're not able to handle a LoadBalancer service type for many reasons. This PR should make it so we can override the service types and set them to NodePort. In our case, we are creating the Ingress controller's external Load Balancer with Terraform, which can also discover the NodePort via the `kubernetes_service` data source.

The expected impact should be the exact same behavior as before since the default service type is LoadBalancer.

Cheers!
Dillon
BOT NOTES: 
resolves https://github.com/solo-io/gloo/pull/2683

### CODE REVIEW

**Comment by kdorosh** [comment]
@iridian-ks thanks for the PR!

We should the new value to our helm `generate` script in this file, similar to: https://github.com/solo-io/gloo/blob/master/install/helm/gloo/generate/values.go#L236

the json annotation description there will be automatically rendered into our docs here: https://docs.solo.io/gloo/latest/reference/helm_chart_values/

(don't forget to run `make generated-code` after updating the generate script before committing)

a helm test as well would be nice, see https://github.com/solo-io/gloo/blob/master/install/test/helm_test.go#L586-L595 for an example

thanks again for the contribution!

**Comment by iridian-ks** [comment]
Hi @kdorosh !

Thanks for the tips. I updated the `values.go` and generated the docs, which worked.

I'm having trouble following the tests as it looks like the existing tests around extraAnnotations are for gatewayProxy? A search for knative came up empty. It does look like the ingressProxy tests might be there but I'm unsure how to add the tests for the Service type as there aren't any existing tests for services.

I could probably figure it out once I find a bit more time. I'll look into it during the week. I could probably create the service tests for knative and ingressProxy. I ran `go test` in the test directory and things look OK.

**Comment by kdorosh** [comment]
Yeah, there are currently no tests for the other service types, but it would be nice if we could add new tests for them.

Also we should add a proper changelog to this PR, see https://github.com/solo-io/go-utils/tree/master/changelogutils for documentation and this changelog file as an example: https://github.com/solo-io/gloo/blob/master/changelog/v1.3.8/extra-proxy-volume-mounts.yaml

**Comment by kdorosh** [comment]
Also, note that ingress service tests will be added as part of this PR, which will probably merge before you look at this PR again: https://github.com/solo-io/gloo/pull/2637

I would pull in those changes to your branch once it merges

**Comment by kdorosh** [comment]
We released version v1.3.16, please update your changelog to v1.4.0-beta1

File: Makefile [hunk 1]
```diff
@@ -414,8 +414,9 @@ HELM_DIR := install/helm/gloo
 .PHONY: generate-helm-files
 generate-helm-files: $(OUTPUT_DIR)/.helm-prepared
 
-$(OUTPUT_DIR)/.helm-prepared:
-	GO111MODULE=on go run $(HELM_DIR)/generate.go --version $(VERSION)  --generate-helm-docs
+HELM_PREPARED_INPUT := $(HELM_DIR)/generate.go $(wildcard $(HELM_DIR)/generate/*.go)
```
Comment (iridian-ks) [inline Makefile:417]: I added this to the Makefile so when you change any of the Go code then Make knows to update the docs. Prior to this the docs wouldn't re-create if the Go code changed. Obviously there's -B, but this seems better.

File: install/helm/gloo/generate/values.go [hunk 2]
```diff
@@ -285,14 +286,10 @@ type IngressDeployment struct {
 }
 
 type IngressProxy struct {
-	Service    *IngressProxyService    `json:"service,omitempty"`
 	Deployment *IngressProxyDeployment `json:"deployment,omitempty"`
 	ConfigMap  *IngressProxyConfigMap  `json:"configMap,omitempty"`
 	Tracing    *string                 `json:"tracing,omitempty"`
-}
-
-type IngressProxyService struct {
```
Comment (iridian-ks) [inline install/helm/gloo/generate/values.go:294]: LMK if this change is OK, there's some duplication between Service's so I made that reusable by bringing ExtraAnnotations into the knative stuff as well.

File: install/test/helm_test.go [hunk 3]
```diff
@@ -1935,6 +1935,16 @@ metadata:
 					testManifest.ExpectService(ingressProxyService)
 				})
 
+				It("sets type", func() {
```
Comment (iridian-ks) [inline install/test/helm_test.go:1938]: This test is failing and I'm unsure why. Maybe someone can explain it to me?

```
Helm Test
/Users/dgilmor/Code/oss/gloo/install/test/helm_test.go:84
  Helm 3
  /Users/dgilmor/Code/oss/gloo/install/test/helm_test.go:131
    ingress-proxy service
    /Users/dgilmor/Code/oss/gloo/install/test/helm_test.go:1889
      sets type [It]
      /Users/dgilmor/Code/oss/gloo/install/test/helm_test.go:1938

      can't find object Service gloo-system ingress-proxy
```

**Comment by iridian-ks** [comment]
@kdorosh I pushed an update and annotated someone the changes I made and issues I'm having. I need a tad bit of help if you have the time.

Thanks!!

File: install/test/helm_test.go (hunk 3 above)
Comment (kdorosh) [reply install/test/helm_test.go:1938]: this means that the rendered helm chart didn't have a Service named ingress-proxy in gloo-system. perhaps the helm values passed in are wrong?

Comment (iridian-ks) [reply install/test/helm_test.go:1938]: Ah, didn't enable the ingressProxy in the values. Looks like all tests are passing now... thanks for the pointer!  :D

**Comment by kdorosh** [comment]
@iridian-ks , don't forget to add a proper changelog file to this PR

https://github.com/solo-io/gloo/pull/2683#issuecomment-605742736

**Pull request merged**